### ISSUE TITLE
Agent silently truncates multi-step tool tasks: qwen3.6-35b-a3b-mxfp4-mtp returns empty (premature EOS) at large context, loop ends on the no-tool-call turn

### ISSUE THREAD

**Issue by jax-0n-git** [osaurus-ai/osaurus#1680]
### Checks
- [x] I can reproduce this with the latest release (0.20.8)
- [x] I have searched existing issues

### Environment
- Osaurus 0.20.8, macOS 26.5 (Tahoe), Apple M5 Max (128 GB)
- Model: `qwen3.6-35b-a3b-mxfp4-mtp` (bundled OsaurusAI MoE/MTP), used as a **custom agent** (per-agent model override)
- Resolved tool-call parser for this model: `"qwen"` (family fallback in `OpenAIAPI.swift`)

### Summary
Running a multi-step **tool-using** task through the in-app **Agent** loop on the local MTP model, the agent reliably **stops before finishing**. It runs the first few tool calls fine, then emits a short tool-action *preamble* (e.g. `"Let me search"`, `"I'll continue searching for…"`) **with no accompanying tool call**, and the turn ends — no summary, no error, no crash. The HTTP request returns **200**; the task is just left incomplete. (Matches user reports of "incomplete responses.")

### Root cause: large context → premature EOS (model returns empty)
The trigger is **context size**, reproducible on the bare `/v1/chat/completions` API with `search`+`fetch` tools defined, continuing right after a tool result:

| context | result |
|---|---|
| ~1.5K tokens | **32/32 CLEAN tool_calls** (8× temp 1.0 + 8× temp 0.0; also 16/16 clean *with* the injected meta-tools `complete`/`clarify`/`capabilities_discover`/`capabilities_load`/`todo` + an agent system prompt) |
| ~60K tokens (`prompt_tokens=60299`) | **empty response** — `finish_reason=stop`, `content=""`, no `tool_calls` (observed in 2/2 completed API runs at temp 1.0; I did not complete a temp=0 large-context arm) |

Once context is large, the model emits a premature/immediate EOS. **The same failure reproduced reliably through the actual in-app Agent loop across 4 turns** (temp 0 and 1, at ~41K and ~81K context): there the model emits a short preamble (e.g. *"Let me search"*) then stops with no tool call; via the bare API at 60K it returns fully empty. Either way the turn has no tool call, and **Osaurus's agent loop treats any assistant turn that lacks a tool call as the final answer and ends the turn** → silent truncation. (6 reproductions total: 4 in-app + 2 API; I have not finely measured the onset threshold — data points at ~1.5K = clean and ~41–60K = failing.) `Max Tool Attempts` (15) is not the cause (truncates after ~3 calls); Insights shows 0 errors / 100% 200-success — a silent content-level failure.

Specificity (not yet cleanly isolated): the ideal same-family control (`qwen3.6-27b-mxfp4`, dense non-MTP) is too slow to test at 60K (>300s just to prefill — bandwidth-bound). The MTP decode path is the prime suspect given the model card's claim that it is bit-identical to plain AR decoding, but a general large-context generation issue on this runtime is not fully excluded. Happy to add a non-MTP comparison if useful.

### What it is NOT (ruled out empirically)
- **Not the tool-call parser / not leak-or-drop** — 32/32 clean `tool_calls` at small context, no `<|tool_call>` markup leaking into content.
- **Not temperature** — clean at temp 1.0 and 0.0; setting the in-app agent to `temperature: 0` did not fix the truncation.
- **Not the meta-tools / agent prompt** — 16/16 clean with them added.
- **Not `capabilities_load`/#1675** — the loop ends on a *text* turn, not a retrying tool error.

### In-app reproduction (reliable)
1. Custom agent on `qwen3.6-35b-a3b-mxfp4-mtp`, auto-discover capabilities ON.
2. Prompt: *"Search the web for recent research on lucid dreaming induction techniques (MILD, WBTB, reality testing) and write a detailed, well-cited summary. Use web search and fetch tools only."*
3. Allow the search/fetch calls. After 2–3 real tool results (context ~40–80K) the turn ends on a bare preamble with no summary. A "continue" nudge just produces another bare preamble.

### Note on expected behavior
The model card (`OsaurusAI/Qwen3.6-35B-A3B-MXFP4-MTP`) documents no large-context limitation and states MTP decode "stays **bit-identical** to plain autoregressive decoding — only faster." So empty output at ~60K is **not** expected per the model authors — pointing at a runtime issue (MTP decode path and/or large-context generation), not a model limitation.

### Suggested direction (either mitigation helps)
- **Loop side (robust):** don't treat an assistant turn as terminal purely because it lacks a tool call when the model is mid-plan (e.g. an open `todo` with unchecked items, or text that announces a next action) — re-prompt/continue instead of ending. (Relates to the completion-semantics work in #825.)
- **Model/runtime side:** investigate why the MTP model emits an empty/premature EOS at ~60K (the MTP decode path is supposed to be bit-identical to plain AR decoding).

### Related
- **#647** ("no results whatsoever", open) — likely the same user-visible symptom without the mechanism; this adds the minimal repro + root cause.
- **#1496** (Gemma weird output with tools) — same symptom family, different model family; this one is Qwen3.6-MoE/MTP and is not parser-leak.
- **#825** (Work Mode reliability: completion semantics) — the relevant loop-termination work.
- **#798** (Gemma crash on 2nd generation at large context) — large-context fragility in another model (crash vs empty).
- **#1228** (Metal SIGABRT) — separate crash; did not reproduce here across many 41–81K runs.

**Comment by jax-0n-git** [comment]
**Follow-up (firmed repro + engine-level root-cause narrowing + confirmed on 0.20.9).** Re-ran on the bare `/v1/chat/completions` API on `127.0.0.1:8770` (`search`+`fetch` tools, continuing right after a tool result), machine cool, model pre-warmed, `max_tokens: 4096`. Also captured the unified-log engine trace, which moves the suspect *off* the MTP draft path.

### 1. Core claim firmed — n=12 at ~60K, both temperatures
`qwen3.6-35b-a3b-mxfp4-mtp`, prompt ≈49K (engine-counted) / 60K (API `usage.prompt_tokens=59752`):
- **temp 1.0: 6/6 broken** — 5× fully empty (`completion_tokens=0`), 1× one token then EOS
- **temp 0.0: 6/6 fully empty** (`completion_tokens=0`)

`completion_tokens=0` + `finish_reason=stop` = the sampler emits EOS as its **first** output token, below the agent loop and below the tool-call parser.

### 2. Onset is context-size dependent and has TWO modes (not only "empty")
Identical filler, only context size varies (temp 1.0, n=4 per size):

| context | result |
|---|---|
| ~20K | **mixed** — 1 clean `todo` tool_call, 1 preamble-then-EOS, 2 empty |
| ~30K | 4/4 empty (`completion_tokens=0`) |
| ~40K | 4/4 empty (`completion_tokens=0`) |
| ~50K | 4/4 **`!!!!` runaway degeneration** to the cap (`finish_reason=length`, 4096 chars of `!`) |
| ~60K | 6/6 empty (`completion_tokens=0`) |

At ~50K it isn't empty — it's token salad. So this issue (empty) and #1496 / #1432 (garbled output) likely share one root cause: a large-context decode corruption surfacing as **either** immediate EOS **or** runaway garbage depending on size.

### 3. Engine trace narrows the cause — it is NOT the MTP draft path
From `log show --predicate 'process == "osaurus"'` during a failing 60K call:
```
[ai.osaurus:BatchAdapter] submit: model=qwen3.6-35b-a3b-mxfp4-mtp promptTokens=48943 temperature=1.0
    topP=0.95 topK=20 minP=0.0 maxTokens=4096 draftStrategy=none nativeMTPFallback=explicit_sampling
[ai.osaurus:Generation] [perf] mlxStats promptTokens=48943 promptTps=402 genTokens=0 genTps=0 stop=stop
[vmlx:TokenIterator] cache hit rolling back to full prefill (path-dependent full cache hit missing seed-boundary SSM state)
```
- **`draftStrategy=none`** at temp 1.0 → the native-MTP *draft/verify* loop is **not active** when this fails, yet `genTokens=0 / stop=stop`. temp 0.0 is also 6/6 empty (I only captured the engine trace for the temp-1.0 runs, so I can't confirm whether MTP draft engaged at temp 0 — but the temp-1.0 leg alone, with draft provably off, already shows the failure). **So the failure does not depend on the MTP draft path** — which I think rules out the issue's current prime suspect.
- **It's bundle-specific, not "hybrid-SSM"-generic.** The dense same-family **`qwen3.6-27b-mxfp4`** (which, per `Qwen35.swift`, is *also* a GatedDeltaNet/SSM hybrid) ran the identical probe **clean at 40K (3/3 `search` tool_calls)** and coherent at 50K (**never `!!!!`**, 1/3 empty). So the catastrophic empty/salad is specific to the `qwen3.6-35b-a3b-mxfp4-mtp` bundle (MoE + MXFP4), at the decode level.
- The `TokenIterator` "missing seed-boundary SSM state" rollback is a real GatedDeltaNet diagnostic, **but I'm not claiming it's the cause** — run-1 (cache *miss*, no rollback) is empty too.

Per the model card, MTP decode is "bit-identical to plain autoregressive decoding" and the context window is 262,144 — so empty/garbled output at 30–60K is far inside spec.

Caveats (kept honest): the *ideal* control — a non-MTP build of the **same** 35B-A3B MoE — isn't bundled, so I can't fully separate "MoE/MXFP4 bundle" from "MTP weights present"; and dense 27B at 60K exceeded a 900s prefill timeout, so there's no 60K control point (40K + 50K already discriminate).

### 4. Loop-side mechanism (precise, against the 0.20.8 *and* 0.20.9 tags)
`Services/Chat/AgentToolLoop.swift` + `Views/Chat/ChatView.swift`: the loop does **not** end on *any* no-tool-call turn. `ChatView` (`invocations.isEmpty`) splits:
- **truly empty** (content blank AND thinking blank) → `.emptyResponse` → driver nudge-retries up to `maxEmptyTurnRetries = 2`, then emits a fallback message (recovered, not a silent dead-end).
- **text present but no tool call** → `.finalResponse` → ends immediately, no retry.

So the in-app silent truncation is specifically the **preamble** (`.finalResponse`) case (e.g. at ~20K the model emits *"To write the best possible answer to ..."* then EOS, taken as the final answer). For the empty case (most of 30–60K), the empty-turn recovery *does* fire — but the model re-emits empty/garbage on each retry, so the user still ends with the fallback and no real result. (Relates to the completion-semantics work in #825.)

### 5. Confirmed on 0.20.9
The `0.20.8...0.20.9` diff touches no engine/MTP files (no `vmlx-swift` / `Package.resolved` bump) and not `AgentToolLoop.swift`, so it's expected to persist — and it does: **60K MTP, temp 1.0 → 4/4 empty on 0.20.9**.

Happy to share the probe harness + raw outputs.

**Comment by jax-0n-git** [comment]
FYI — not asking to reopen (the loop-side surface fix in #1681 is the right call given what's below). I dug the engine side of this on an M5 Max / macOS 26.5.1 and have a differential that may save time if the engine path is ever revisited.

**Same model, two MLX engines — only one corrupts:**
- `OsaurusAI/Qwen3.6-35B-A3B-MXFP4-MTP` (`qwen3_5_moe`, 256-expert top-8, **affine int4 g32**) on Osaurus / `mlx-swift` (the v0.31.1 pin): **dead-ends at ~40K context** — empty turn / immediate EOS at temp 0 (3/3), and at temp 1 the logits are finite + EOS-biased (recovers a valid turn ~1/3). Clean below ~20K.
- The **same weights** served via Python **`mlx-vlm` 0.5.0**: **clean to 81K prompt tokens** (full multi-step tool output, temp 0).
- Same-family **dense** `qwen3.6-27b-mxfp4` on `mlx-swift`: **clean at 40K**.

So it isn't the model, and it isn't "4-bit MoE can't do long context" — it's specific to the `mlx-swift` MoE long-context path.

**Ruled out** (source review + live probe): it is **not** the Neural Accelerator kernels — `MLX_METAL_NO_NAX=1` reproduces it identically — and the affine gather kernels (`affine_gather_qmm_t/rhs`, vanilla and `_nax`) are code-correct (fp32 accumulators, 64-bit offsets, correct `dequantize_scale`).

**Surviving suspects:** the mlx version pinned in `mlx-swift` vs the one `mlx-vlm` bundles, KV-cache precision, or the routed-expert (gather) prefill path. Happy to run a targeted repro / bisect on the M5 if it'd help.
